discovery_engine_core/core/src/stack/semantic_filters.rs

```diff
@@ -0,0 +1,158 @@
+#[allow(dead_code)]
+fn determine_semantic_clusters(
+    documents: &[Document],
+    method: Method,
+    distance_threshold: f32,
+) -> Vec<usize> {
+    if documents.len() < 2 {
+        return vec![0; documents.len()];
+    }
+
+    let mut condensed_dissimilarity_matrix = condensed_cosine_distance(documents);
+    debug_assert_eq!(
+        condensed_dissimilarity_matrix.len(),
+        (documents.len() * (documents.len() - 1)) / 2,
+    );
+
+    let dendrogram = linkage(&mut condensed_dissimilarity_matrix, documents.len(), method);
+    cut_tree(&dendrogram, distance_threshold)
+}
+
+/// Computes the condensed cosine distance matrix of the documents' embeddings.
+fn condensed_cosine_distance(documents: &[Document]) -> Vec<f32> {
+    pairwise_cosine_similarity(
+        documents
+            .iter()
+            .map(|document| document.smbert_embedding.view()),
+    )
+    .indexed_iter()
+    .filter_map(|((i, j), &similarity)| (i < j).then(|| 1. - similarity))
+    .collect()
+}
+
+/// Cuts off the dendrogram at the first step which exceeds the distance/dissimilarity threshold.
+fn cut_tree(dendrogram: &Dendrogram<f32>, threshold: f32) -> Vec<usize> {
+    // at the beginning every sample is in its own cluster
+    let clusters = (0..dendrogram.observations())
+        .map(|x| (x, vec![x]))
+        .collect::<BTreeMap<_, _>>();
+
+    // merge clusters until threshold is reached
+    let (_, clusters) = dendrogram
+        .steps()
+        .iter()
+        .take_while(|step| step.dissimilarity < threshold)
+        .fold(
+            (dendrogram.observations(), clusters),
+            |(id, mut clusters), step| {
+                // unwrap safety:
+                // - inital cluster ids have been inserted in the beginning
+                // - merged cluster ids have been inserted in a previous iteration/step
+                let mut cluster_1 = clusters.remove(&step.cluster1).unwrap();
+                let cluster_2 = clusters.remove(&step.cluster2).unwrap();
+
+                // merge clusters
+                cluster_1.extend(cluster_2);
+                clusters.insert(id, cluster_1);
+
+                (id + 1, clusters)
+            },
+        );
+
+    // assign labels to samples
+    clusters.into_values().enumerate().fold(
+        vec![0; dendrogram.observations()],
+        |mut labels, (label, cluster)| {
+            for sample in cluster {
+                labels[sample] = label;
+            }
+            labels
+        },
+    )
+}
+
+#[cfg(test)]
+#[allow(clippy::non_ascii_literal)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_cluster_empty_documents() {
+        let labels = determine_semantic_clusters(&[], Method::Average, 1.);
+        assert!(labels.is_empty());
+    }
+
+    #[test]
+    fn test_cluster_single_document() {
+        let labels = determine_semantic_clusters(&[Document::default()], Method::Average, 1.);
+        assert_eq!(labels, [0]);
```

Comment:
Here we deviate slightly from the `sklearn` implementation. `sklearn` always requires at least two samples but I think that shouldn't be a problem for us.

Reply:
i think so too. for sklearn one would need additional error handling for the case < 2, here we just baked that into the function with a trivial extension.